fl4health/feature_alignment/tab_features_info_encoder.py

```diff
@@ -0,0 +1,194 @@
+import json
+from typing import Dict, List
+
+import pandas as pd
+from cyclops.process.feature.feature import TabularFeatures
+from flwr.common.typing import Scalar
+from sklearn.feature_extraction.text import CountVectorizer
+
+from fl4health.feature_alignment.constants import BINARY, DEFAULT_FILL_VALUES, FEATURE_TYPES, ORDINAL, STRING
+from fl4health.feature_alignment.string_columns_transformer import StringColumnTransformer
+
+
+class TargetInfoEncoder:
+    """
+    This class encodes the information about the target column
+    that is necessary to perform feature alignment.
+
+    Parameters
+    ----------
+    target: str
+        Target column.
+    target_type: str
+        Target column data type. Must be one of BINARY, NUMERICAL, ORDINAL, and STRING.
+    target_categories: List[Scalar]
+        The categories of the target column, provided it is of type ORDINAL.
+    """
+
+    def __init__(self, target: str, target_type: str, target_categories: List[Scalar]) -> None:
+        self.target = target
+        self.target_type = target_type
+        self.target_categories = target_categories
+
+    def get_target(self) -> str:
+        return self.target
+
+    def get_target_type(self) -> str:
+        return self.target_type
+
+    def get_target_categories(self) -> List[Scalar]:
+        return self.target_categories
+
+    def to_json(self) -> str:
+        return json.dumps(
+            {
+                "target": json.dumps(self.get_target()),
+                "target_type": json.dumps(self.get_target_type()),
+                "target_categories": json.dumps(self.get_target_categories()),
+            }
+        )
+
+    @staticmethod
+    def from_json(json_str: str) -> "TargetInfoEncoder":
+        attributes = json.loads(json_str)
+        return TargetInfoEncoder(
+            json.loads(attributes["target"]),
+            json.loads(attributes["target_type"]),
+            json.loads(attributes["target_categories"]),
+        )
+
+
+class TabularFeaturesInfoEncoder:
+    """
+    This class encodes all the information required to perform feature
+    alignment on tabular datasets.
+
+    Parameters
+    ----------
+    features_to_types: Dict[str, str]
+        Dictionary that maps each feature name to its type.
+        We consider four types in tabular data:
+        BINARY, ORDINAL, NUMERICAL, and STRING.
+    categories: Dict[str, List[Scalar]]
+        Dictionary that maps each ordinal feature to its categories.
+    target_info: TargetInfoEncoder
+        Information about the target column(s).
+    vocabulary: Dict[str, int]
+        Vocabulary of all the STRING columns.
```

Comment:
I think, by default, we want to have a vocabulary per STRING column rather than one for all columns.